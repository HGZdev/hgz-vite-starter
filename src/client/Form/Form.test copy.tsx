// import {describe, it, expect, vi} from "vitest";
// import {render, fireEvent} from "@testing-library/react";

// // Mock function for onSubmit
// const onSubmitMock = vi.fn();

// // Test suite for Input component
// describe("Input Component Tests", () => {
//   // Test case for rendering Input component
//   it("should render Input component with label", () => {
//     const {container} = render(
//       <Input
//         type="text"
//         id="username"
//         value=""
//         onChang// FormikForm.test.tsx
// import { describe, it, expect, setup } from 'vitest';
// import { render, fireEvent } from '@testing-library/react';
// import { FormikForm } from './FormikForm';

// describe('FormikForm', () => {
//   it('renders and submits FormikForm', async () => {
//     const handleSubmit = setup.spy();
//     const user = setup.user();

//     render(
//       <FormikForm onSubmit={handleSubmit}>
//         <input type="text" name="field1" />
//         <button type="submit">Submit</button>
//       </FormikForm>
//     );

//     await user.type('input[name="field1"]', 'TestValue');
//     await user.click('button[type="submit"]');

//     await setup.waitFor(() => {
//       expect(handleSubmit).toHaveBeenCalledWith({ field1: 'TestValue' });
//     });
//   });
// });

// // InputField.test.tsx
// import { describe, it, expect } from 'vitest';
// import { render } from '@testing-library/react';
// import { InputField } from './InputField';

// describe('InputField', () => {
//   it('renders InputField', () => {
//     render(<InputField label="Test Label" name="testField" type="text" />);

//     // Check if the InputField is rendered with the provided label
//     expect(document.querySelector('label[for="testField"]')).not.toBeNull();
//   });
// });
// e={() => {}}
//         label="Username"
//         required
//       />
//     );

//     // Example assertion using vitest
//     expect(container.querySelector('label[for="username"]')).toBeTruthy();
//   });

//   // Test case for input change
//   it("should call onChange when input value changes", () => {
//     const onChangeMock = vi.fn();
//     const {getByLabelText} = render(
//       <Input
//         type="text"
//         id="username"
//         value=""
//         onChange={onChangeMock}
//         label="Username"
//         required
//       />
//     );

//     const inputElement = getByLabelText("Username");
//     fireEvent.change(inputElement, {target: {value: "USERNAME"}});

//     expect(onChangeMock).toHaveBeenCalledTimes(1);
//     expect(onChangeMock).toHaveBeenCalledWith(
//       expect.objectContaining({target: {value: "USERNAME"}})
//     );
//   });
// });

// // Test suite for Form component
// describe("Form Component Tests", () => {
//   // Test case for rendering Form component
//   it("should render Form component with onSubmit handler", () => {
//     const {container} = render(
//       <Form onSubmit={onSubmitMock}>
//         <Input
//           type="text"
//           id="username"
//           value=""
//           onChange={() => {}}
//           label="Username"
//           required
//         />
//         {/* Add other necessary children components */}
//         <SubmitButton type="submit">Submit</SubmitButton>
//       </Form>
//     );

//     // Example assertion using vitest
//     expect(container.querySelector("form")).toBeTruthy();
//   });

//   // Test case for form submission
//   it("should call onSubmit when the form is submitted", () => {
//     const {container} = render(
//       <Form onSubmit={onSubmitMock}>
//         <Input
//           type="text"
//           id="username"
//           value=""
//           onChange={() => {}}
//           label="Username"
//           required
//         />
//         {/* Add other necessary children components */}
//         <SubmitButton type="submit">Submit</SubmitButton>
//       </Form>
//     );

//     const formElement = container.querySelector("form") as HTMLFormElement;

//     fireEvent.submit(formElement);
//     expect(onSubmitMock).toHaveBeenCalledTimes(1);
//   });

//   // Test case for rendering Form component with children
//   it("should render Form component with children", () => {
//     const {getByLabelText, getByText} = render(
//       <Form onSubmit={onSubmitMock}>
//         <Input
//           type="text"
//           id="username"
//           value=""
//           onChange={() => {}}
//           label="Username"
//           required
//         />
//         <Input
//           type="password"
//           id="password"
//           value=""
//           onChange={() => {}}
//           label="Password"
//           required
//         />
//         <SubmitButton type="submit">Submit</SubmitButton>
//       </Form>
//     );

//     expect(getByLabelText("Username")).toBeTruthy();
//     expect(getByLabelText("Password")).toBeTruthy();
//     expect(getByText("Submit")).toBeTruthy();
//   });

//   // Test case for handling form submission with validation
//   it("should not call onSubmit when form is submitted with invalid fields", () => {
//     const {getByText, getByLabelText} = render(
//       <Form onSubmit={onSubmitMock}>
//         <Input
//           type="text"
//           id="username"
//           value=""
//           onChange={() => {}}
//           label="Username"
//           required
//           validate={["email"]}
//         />
//         <SubmitButton type="submit">Submit</SubmitButton>
//       </Form>
//     );

//     const formElement = getByText("Submit").closest("form") as HTMLFormElement;
//     fireEvent.submit(formElement);
//     // Assert that onSubmitMock has not been called
//     expect(onSubmitMock).not.toHaveBeenCalled();

//     const errorMessage = getByLabelText("Username").nextSibling;
//     expect(errorMessage).toEqual("Please enter required field(s)");
//   });
// });
