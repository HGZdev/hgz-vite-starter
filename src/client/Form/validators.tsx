const emailRegex =
  // eslint-disable-next-line no-control-regex
  /^(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])$/;

const passwordRegex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d]{8,}/; //  It requires a password to have at least one lowercase letter, one uppercase letter, one digit, and be at least 8 characters long.

type Validators = Record<string, (value: string) => string | null>;

export const validators: Validators = {
  required: (v: string) =>
    !(v && v.length > 0) ? "Please enter required field(s)" : null,
  email: (v: string) =>
    !emailRegex.test(v) ? "Please enter a valid email address" : null,
  password: (v: string) =>
    !passwordRegex.test(v)
      ? "Password must be at least 8 characters long and contain at least one uppercase letter, one lowercase letter, and one number"
      : null,
};
