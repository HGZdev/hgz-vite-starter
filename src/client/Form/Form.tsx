import React, {
  ChangeEvent,
  FC,
  ReactNode,
  useState,
  FocusEventHandler,
  useEffect,
} from "react";
import styled from "styled-components";
import {validators as defaultValidators, Validators} from "./validators";

// Styled Components
const FormContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const StyledForm = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const Label = styled.label`
  margin-bottom: 0.5rem;
`;

const StyledInput = styled.input`
  margin-bottom: 1rem;
  padding: 0.5rem;
`;

export const SubmitButton = styled.button`
  padding: 0.5rem 1rem;
`;

export const ErrorLabel = styled.span`
  color: red;
`;

export type InputValue = string | number | readonly string[] | undefined;

type FieldValidationProps = {
  value?: InputValue;
  validators?: Validators;
  validate?: string[];
  required?: boolean;
};

const makeValidation = ({
  value = undefined,
  validators = defaultValidators,
  validate = [],
  required,
}: FieldValidationProps) => {
  let fieldError;

  const rules = required ? [...(validate || []), "required"] : validate;

  for (const rule of rules || []) {
    const validationError = validators[rule](value) ?? undefined;
    fieldError = validationError;
    if (validationError) break;
  }

  return fieldError;
};

interface InputProps {
  type: string;
  id: string;
  value?: InputValue;
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  label?: string;
  required?: boolean;
  ariaLabel?: string;
  validate?: string[];
  autoComplete?: string;
}

// Input Component
export const Input: FC<InputProps> = ({
  type,
  id,
  value,
  onChange,
  onBlur,
  required,
  validate,
  label = id,
  ariaLabel = label || id,
  autoComplete,
}) => {
  const [fieldError, setFieldError] = useState<string>();

  const handleChange = (e: React.FocusEvent<HTMLInputElement>) => {
    setFieldError(
      makeValidation({
        value: e.target.value,
        validate,
        required,
      })
    );
    onChange?.(e);
  };

  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    setFieldError(
      makeValidation({
        value: e.target.value,
        validate,
        required,
      })
    );
    onBlur?.(e);
  };

  return (
    <>
      <Label htmlFor={id}>{label}:</Label>
      <StyledInput
        type={type}
        id={id}
        value={value}
        onChange={handleChange}
        onBlur={handleBlur}
        aria-label={ariaLabel}
        required={required}
        autoComplete={autoComplete}
      />
      {fieldError && <ErrorLabel>{fieldError}</ErrorLabel>}
    </>
  );
};

interface FormProps {
  children: ReactNode;
  onSubmit: (values: Record<string, InputValue>) => void;
  onError?: (error: string) => void;
  showSubmitButton?: boolean;
  initialValues?: Record<string, InputValue>;
}

// Form Component
export const Form: FC<FormProps> = ({
  children,
  onSubmit,
  onError,
  showSubmitButton,
  initialValues,
}) => {
  const [formValues, setFormValues] = useState<Record<string, InputValue>>({});

  useEffect(() => {
    console.log("initialValues:", initialValues);
    if (initialValues) setFormValues(initialValues);
  }, [initialValues]);

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    // Check for field validation errors
    const validationErrors: Record<string, string | undefined> = {};
    React.Children.forEach(children, (child) => {
      if (React.isValidElement(child) && child.type === Input) {
        const {id, validate, required} = child.props as InputProps;
        validationErrors[id] = makeValidation({
          value: formValues[id],
          validate,
          required,
        });
      }
    });

    if (Object.values(validationErrors).length > 0) {
      return;
    }

    try {
      // If validation passes, submit the form
      onSubmit(formValues);
    } catch (error) {
      if (onError && error instanceof Error) {
        onError(error.message);
      }
    }
  };

  const handleInputChange = (id: string, value: string) => {
    setFormValues((prevValues) => ({
      ...prevValues,
      [id]: value,
    }));
  };

  return (
    <FormContainer>
      <StyledForm onSubmit={handleSubmit}>
        {React.Children.map(children, (child) => {
          if (React.isValidElement(child) && child.type === Input) {
            const {id} = child.props as InputProps;

            const newInputProps: Partial<InputProps> = {
              onChange: (e: React.ChangeEvent<HTMLInputElement>) =>
                handleInputChange(id, e.target.value),
              onBlur: (e: React.FocusEvent<HTMLInputElement>) => {
                handleInputChange(id, e.target.value);
              },
            };

            return React.cloneElement(child, {
              ...child.props,
              ...newInputProps,
            });
          }
          return child;
        })}
        {showSubmitButton && <button type="submit">Submit</button>}
      </StyledForm>
      <>{JSON.stringify(formValues)}</>
    </FormContainer>
  );
};
