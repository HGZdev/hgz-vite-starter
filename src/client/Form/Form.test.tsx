import {render, screen, waitFor} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {expect, test, vi} from "vitest";
import {FormikForm, InputField} from "./Form";
import * as Yup from "yup";

test("renders and submits the FormikForm component with InputField", async () => {
  const handleSubmit = vi.fn();
  interface UserData {
    firstName: string;
    lastName: string;
  }

  const initialValues = {firstName: "", lastName: ""};

  render(
    <FormikForm<UserData> onSubmit={handleSubmit} initialValues={initialValues}>
      <InputField label="First Name" name="firstName" type="text" />
      <InputField label="Last Name" name="lastName" type="text" />
      <button type="submit">Submit</button>
    </FormikForm>
  );

  const user = userEvent.setup();
  await user.type(screen.getByLabelText(/First Name/i), "John");
  await user.type(screen.getByLabelText(/Last Name/i), "Doe");
  await user.click(screen.getByRole("button", {name: /submit/i}));

  await waitFor(() => {
    expect(handleSubmit).toHaveBeenCalledTimes(1);
    const [submittedValues] = handleSubmit.mock.calls[0];
    expect(submittedValues).toEqual({
      firstName: "John",
      lastName: "Doe",
    });
  });
});

test("renders and handles errors in the FormikForm component with InputField", async () => {
  const handleSubmit = vi.fn();
  interface UserData {
    firstName: string;
    lastName: string;
  }
  const initialValues = {
    firstName: "",
    lastName: "",
  };
  // Define Yup schema for validation
  const validationSchema = Yup.object({
    firstName: Yup.string().required("First Name is required"),
    lastName: Yup.string().required("Last Name is required"),
  });

  render(
    <FormikForm<UserData>
      onSubmit={handleSubmit}
      initialValues={initialValues}
      validationSchema={validationSchema}
    >
      <InputField label="First Name" name="firstName" type="text" />
      <InputField label="Last Name" name="lastName" type="text" />
      <button type="submit">Submit</button>
    </FormikForm>
  );

  const user = userEvent.setup();

  // Simulate error by not providing a value for the required fields
  await user.click(screen.getByRole("button", {name: /submit/i}));

  await waitFor(() => {
    // Ensure handleSubmit is not called since there are errors
    expect(handleSubmit).not.toHaveBeenCalled();

    // Check for error messages
    expect(screen.getByText(/First Name is required/)).toBeTruthy();
    expect(screen.getByText(/Last Name is required/)).toBeTruthy();
  });
});
