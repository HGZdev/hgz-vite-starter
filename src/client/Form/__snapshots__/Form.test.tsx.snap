// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders and submits the FormikForm component with InputField 1`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "firstName": "John",
        "lastName": "Doe",
      },
      {
        "resetForm": [Function],
        "setErrors": [Function],
        "setFieldError": [Function],
        "setFieldTouched": [Function],
        "setFieldValue": [Function],
        "setFormikState": [Function],
        "setStatus": [Function],
        "setSubmitting": [Function],
        "setTouched": [Function],
        "setValues": [Function],
        "submitForm": [Function],
        "validateField": [Function],
        "validateForm": [Function],
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
